src/main/java/com/microsoft/azure/util/AzureCredentials.java

```diff
@@ -368,24 +360,25 @@ public boolean validate() throws ValidationException {
                 TokenCredential credential;
 
                 if (StringUtils.isEmpty(secret)) {
-                    CertificateCredentialsImpl certificate = getCertificate();
+                    StandardCertificateCredentials certificate = getCertificate();
                     if (certificate == null) {
                         throw new ValidationException(Messages.Azure_ClientCertificate_NotFound());
                     }
-                    ByteArrayInputStream certificateBytes = new ByteArrayInputStream(
-                            certificate.getKeyStoreSource().getKeyStoreBytes());
+
+                    byte[] pkcs12Bytes = getPKCS12Bytes(certificate.getKeyStore(), certificate.getPassword());
+                    ByteArrayInputStream certificateBytes = new ByteArrayInputStream(pkcs12Bytes);
 
                     IdentityClientOptions identityClientOptions = new IdentityClientOptions();
                     identityClientOptions.setHttpClient(HttpClientRetriever.get());
 
-                    credential = new ClientCertificateCredential2(
-                            getTenant(),
-                            getClientId(),
-                            null,
-                            // this is package-private in the default sdk method which is why we have our own class
-                            certificateBytes,
-                            certificate.getPassword().getPlainText(),
-                            identityClientOptions);
+                    credential = new ClientCertificateCredentialBuilder()
+                            .authorityHost(profile.getEnvironment().getActiveDirectoryEndpoint())
+                            .clientId(getClientId())
+                            .pemCertificate(certificateBytes)
```

Comment:
PEM != PKCS12/PFX... ?

Reply:
if you want the PEM then you can pass the `java.security.cert.Certificate` to BouncyCastle-API's PEMEncodeable.
or use `ClientCertificateCredentialBuilder`  (If you just grab the Certificate from the KeyStore and go via PEMEncodeable you will not need to do keystore conversion.

Reply:
The API for pem and pfx is identical under the hood for an input stream. Doesn’t matter which one I call. I’ll check that

Reply:
> if you want the PEM then you can pull the `java.security.cert.Certificate` from the keystore and pass it to BouncyCastle-API's PEMEncodeable.


You may need to also extract the key and chain (not entirely sure with the pem encoding off the top of my head)

Reply:
I've tried the `PEMEncodeable` approach and while I can get the key and pass it to the library I get an error:

```
java.lang.IllegalArgumentException: PEM certificate provided does not contain -----BEGIN CERTIFICATE-----END CERTIFICATE----- block
```

The returned content from:

```
            return PEMEncodable.create(ks.getKey(ks.aliases().nextElement(), plainTextPassword.toCharArray()))
                    .encode();
```

includes `BEGIN RSA PRIVATE KEY` meaning its `PKCS#1 formatted.

I've tried a bunch of other options and haven't found anything that works 😢

Reply:
Can you share a test file that I can upload for testing (or what would be uploaded as the Jenkins credentials)?
Just want to check that anything I generate would be entirely representative.  Will try and find some time to investigate as being able to export is a valid use case that should still be possible.

Reply:
I was able to generate a pkcs#8 formatted one with:

```
            JcaPKCS8Generator gen1 = new JcaPKCS8Generator((PrivateKey) key, null);
            PemObject obj1 = gen1.generate();
            StringWriter sw1 = new StringWriter();
            try (JcaPEMWriter pw = new JcaPEMWriter(sw1)) {
                pw.writeObject(obj1);
            }

            return sw1.toString().getBytes();
```

but that just fails because the certificate is missing which I haven't been able to retrieve, pulling from the keystore its null

Reply:
Attached an example file generated in the same way (password is an empty string, which is the standard way for Azure PFX files...)

[timjenkins-test-for-james-20241205.pfx.zip](https://github.com/user-attachments/files/18030817/timjenkins-test-for-james-20241205.pfx.zip)

Reply:
```
keytool -list -keystore timjenkins-vm-agents-20241201.pfx -alias 5308f259-27a0-42cc-abbe-79b863260c67 -storepass "" -v
```

```
Alias name: 5308f259-27a0-42cc-abbe-79b863260c67
Creation date: 7 Dec 2024
Entry type: PrivateKeyEntry
Certificate chain length: 1
Certificate[1]:
Owner: CN=VM Agents
Issuer: CN=VM Agents
Serial number: 3f24a57e0a054fc9b19f737b20212e1e
Valid from: Sun Dec 01 21:41:08 GMT 2024 until: Fri Dec 01 21:51:08 GMT 2034
Certificate fingerprints:
	 SHA1: AE:4A:A0:D8:9E:65:4B:1D:56:72:1B:07:A6:43:AA:2C:6F:B7:82:CD
	 SHA256: A4:05:D2:DF:83:E1:17:0B:81:FA:AE:7A:6C:4E:2F:B6:BC:E9:44:8B:95:3D:A8:06:AE:AA:44:7C:3C:AA:39:A1
Signature algorithm name: SHA256withRSA
Subject Public Key Algorithm: 2048-bit RSA key
Version: 3
...
```

When I store it the certificate is missing:

```java
ByteArrayOutputStream out = new ByteArrayOutputStream();
ks.store(out, plainTextPassword.toCharArray());

try(OutputStream outputStream = new FileOutputStream("/tmp/keystore.pfx")) {
    out.writeTo(outputStream);
}
```

```
❯ keytool -list -keystore /tmp/keystore.pfx -alias 5308f259-27a0-42cc-abbe-79b863260c67 -storepass "" -v
Alias name: 5308f259-27a0-42cc-abbe-79b863260c67
Creation date: 7 Dec 2024
Entry type: PrivateKeyEntry
Certificate chain length: 0
```

Reply:
I think it will be fixed after: https://github.com/jenkinsci/credentials-plugin/pull/581

will test soon although https://github.com/jenkinsci/plugin-pom/issues/1064 makes this annoying

Reply:
Confirmed works after https://github.com/jenkinsci/credentials-plugin/pull/581 and https://github.com/jenkinsci/azure-credentials-plugin/pull/274/commits/495076b159d442baa79a194f3db0d6345876260e

I still need to test pem

Reply:
Pem was fine